### PR TITLE
Fix sample writer missing thread begin events

### PR DESCRIPTION
**Pull request by phijor**
For each thread location, thread begin events were written lazily in
conjunction with the first recorded  sample.  If there weren't any
samples recorded for this location, the sample writer would 'forget'
to write the thread begin events.

Now, the sample writer will detect a missing thread begin event before
writing the corresponding thread end event, and write the former with
the earliest timestamp available.

### CODE REVIEW

File: src/perf/sample/writer.cpp [hunk 1]
```diff
@@ -208,9 +208,26 @@ void Writer::end()
...
         // time::now() can sometimes can be in the past :-(
-        last_time_point_ = std::max(last_time_point_, lo2s::time::now());
+        auto now = lo2s::time::now();
+        last_time_point_ = std::max(last_time_point_, now);
+
+        // If we have never written any samples on this location, we also never
+        // got to write the thread_begin event.  Make sure we do that now.
+        if (first_event_)
+        {
+            first_time_point_ = std::min(first_time_point_, now);
```
Comment (bmario) [inline src/perf/sample/writer.cpp]:
```suggestion
```

This check is pointless. `first_time_point_` will always be smaller (or equal -- veeeeeeery unlikely) than `now`. lo2s::time::now() is a steady clock. We discussed that already in the other PR.

File: src/perf/sample/writer.cpp [hunk 2]
```diff
@@ -208,9 +208,26 @@ void Writer::end()
...
+            otf2_writer_ << otf2::event::thread_begin(first_time_point_, trace_.process_comm(pid_),
+                                                      -1);
+
+            // Make sure that our last timepoint is *not* before the first
+            // timepoint.  This way, events on this location span a non-negative
+            // ammount of time.
+            last_time_point_ = std::max(last_time_point_, first_time_point_);
```
Comment (bmario) [inline src/perf/sample/writer.cpp]:
```suggestion      
```

This check is also not required. A few lines up, we make sure that `last_time_point_ >= now` and `now >= _first_time_point_`. Therefore, here `last_time_point_` is already greater (or equal) to `first_time_point_`.

File: src/perf/sample/writer.cpp (hunk 1 above)
Comment (phijor) [reply src/perf/sample/writer.cpp]: *mea culpa.*

File: src/perf/sample/writer.cpp (hunk 2 above)
Comment (phijor) [reply src/perf/sample/writer.cpp]: Transitive properties are hard.

File: src/perf/sample/writer.cpp [hunk 3]
```diff
@@ -208,9 +208,26 @@ void Writer::end()
...
+        // got to write the thread_begin event.  Make sure we do that now.
+        if (first_event_)
+        {
+            first_time_point_ = std::min(first_time_point_, now);
+            otf2_writer_ << otf2::event::thread_begin(first_time_point_, trace_.process_comm(pid_),
+                                                      -1);
+
+            // Make sure that our last timepoint is *not* before the first
+            // timepoint.  This way, events on this location span a non-negative
+            // ammount of time.
```
[the remark this answers is not in the document]
Comment (phijor) [reply src/perf/sample/writer.cpp]: I'm not even sure how this got into *a second* commit.

File: src/perf/sample/writer.cpp (hunk 1 above)
Comment (bmario) [reply src/perf/sample/writer.cpp]: I'm confused 😕 It's still in?

**Pull request merged**